File: MapVisualizations/chart.py
from flask import Flask, render_template, jsonify
import altair as alt
import pandas as pd

app = Flask(__name__)

# Load the CSV data
monthly_data = pd.read_csv('static/monthly_stats.csv')

def create_chart(selected_year):
    # Filter data for the selected year
    monthly_filtered_data = monthly_data[monthly_data['Year'] == selected_year]

    # Total Fires by Month
    chart_fires = alt.Chart(monthly_filtered_data).mark_bar(color='brown').encode(
        x=alt.X('Month:O', title='Month #'),
        y=alt.Y('total_fires:Q', title='Total Fires')
    ).properties(
        title=f'Total Fires Each Month in {selected_year}'
    )

    # Total Acres Burned by Month
    chart_acres = alt.Chart(monthly_filtered_data).mark_bar(color='black').encode(
        x=alt.X('Month:O', title='Month #'),
        y=alt.Y('total_acres:Q', title='Total Acres Burned')
    ).properties(
        title=f'Total Acres Burned by Month in {selected_year}'
    )

    # Median Fire Duration by Month
    chart_duration = alt.Chart(monthly_filtered_data).mark_bar(color='orange').encode(
        x=alt.X('Month:O', title='Month #'),
        y=alt.Y('median_monthly_duration:Q', title='Median Fire Duration')
    ).properties(
        title=f'Median Fire Duration in {selected_year}'
    )

    # Total Deaths by Month
    chart_deaths = alt.Chart(monthly_filtered_data).mark_bar(color='red').encode(
        x=alt.X('Month:O', title='Month #'),
        y=alt.Y('total_deaths:Q', title='Total Deaths')
    ).properties(
        title=f'Total Deaths by Month in {selected_year}'
    )

    return chart_fires, chart_acres, chart_duration, chart_deaths

@app.route('/')
def index():
    return render_template('chartindex.html')

@app.route('/get_chart/<int:selected_year>', methods=['GET'])
def get_chart(selected_year):
    chart_fires, chart_acres, chart_duration, chart_deaths = create_chart(selected_year)
    
    # Convert charts to JSON to be rendered with Altair's vega-lite library in the frontend
    return jsonify({
        'chart_fires': chart_fires.to_json(),
        'chart_acres': chart_acres.to_json(),
        'chart_duration': chart_duration.to_json(),
        'chart_deaths': chart_deaths.to_json()
    })

if __name__ == '__main__':
    app.run(debug=True)

File: MapVisualizations/static/chart.js
document.getElementById('year-dropdown').addEventListener('change', function() {
    const selectedYear = this.value;
    fetch(`/get_chart/${selectedYear}`)
        .then(response => response.json())
        .then(data => {
            vegaEmbed('#chart-fires', data.chart_fires);
            vegaEmbed('#chart-acres', data.chart_acres);
            vegaEmbed('#chart-duration', data.chart_duration);
            vegaEmbed('#chart-deaths', data.chart_deaths);

            // Update the total stats for the selected year
            updateCharts(data, selectedYear);
        })
        .catch(error => console.error('Error fetching chart data:', error));
});

// Trigger the first load
document.getElementById('year-dropdown').dispatchEvent(new Event('change'));

// Function to update the charts based on the selected year
function updateCharts(data, selectedYear) {
    // Filter data by the selected year
    const filteredData = data.filter(d => d.Year == selectedYear)[0];

    // Console log the filtered data to verify results
    console.log("Filtered Data:", filteredData);

    // Update key stats for the selected year
    const totalStatsDiv = d3.select("#total-stats");
    totalStatsDiv.html(`
        <p><strong>Total Fires This Year:</strong> ${filteredData.total_fires}</p>
        <p><strong>Total Acres Burned:</strong> ${parseFloat(filteredData.total_acres).toLocaleString('en-US', { minimumFractionDigits: 0, maximumFractionDigits: 0 })}</p>
        <p><strong>Median Fire Duration:</strong> ${filteredData.median_yearly_duration}</p>
        <p><strong>Total Deaths:</strong> ${filteredData.total_deaths}</p>
        <p><strong>Total Financial Damage, $:</strong> ${parseFloat(filteredData.Tot_Damage).toLocaleString('en-US', { minimumFractionDigits: 0, maximumFractionDigits: 0 })}</p>
    `);
}
